rs/sns/governance/src/governance.rs

```diff
@@ -477,6 +479,51 @@ impl GovernanceProto {
                 )
             })
     }
+
+    pub fn custom_functions_to_topics(&self) -> BTreeMap<u64, Option<Topic>> {
+        self.id_to_nervous_system_functions
+            .iter()
+            .filter_map(|(function_id, function)| {
+                match &function.function_type {
+                    Some(FunctionType::GenericNervousSystemFunction(generic)) => {
+                        if let Some(topic) = generic.topic {
+                            let specific_topic = match Topic::try_from(topic) {
+                                Err(err) => {
+                                    log!(
+                                        ERROR,
+                                        "Custom proposal ID {function_id}: Cannot interpret \
+                                            {topic} as Topic: {err}",
+                                    );
+
+                                    // This should never happen; if it somehow does, treat this
+                                    // case as a custom function for which the topic is unknown.
+                                    None
+                                }
+                                Ok(Topic::Unspecified) => {
+                                    log!(
+                                        ERROR,
+                                        "Custom proposal ID {function_id}: topic Unspecified."
+                                    );
+
+                                    // This should never happen, but if it somehow does, treat this
+                                    // case as a custom function for which the topic is unknown.
+                                    None
+                                }
+                                Ok(topic) => Some(topic),
+                            };
+
+                            Some((*function_id, specific_topic))
+                        } else {
+                            // Topic not yet set for this custom function.
+                            Some((*function_id, None))
+                        }
+                    }
+                    // Not a custom function.
+                    _ => None,
```

Comment:
```suggestion
                // Skip native proposals.
                let Some(FunctionType::GenericNervousSystemFunction(function) = &function.function_type else {
                    return None;
                };
                
                // No need to convert topic if function has no topic.
                let Some(topic) = function.topic else {
                    return Some((*function_id, None));
                }
                
                // Convert topic from i32 to enum.
                let mut topic = Topic::try_from(topic)
                    .inspect_err(|err| {
                        // log
                    })
                    .ok(); // Convert from Result to Option.
                    
                // Treat Unspecified the same as no topic (and normalize).
                if topic == Some(Topic::Unspecified) {
                    // log
                    topic = None;
                }
                
                Some((*function_id, topic))
```

My opinion: https://www.tumblr.com/magic-bad/764948183732387840/nesting-bad

Reply:
Great suggestion thanks!